content/blog/2019-10-09-loops/index.md

```diff
@@ -0,0 +1,235 @@
+once in the second. We can speed up this loop by fusing the loops. This way,
+we load a cache line and perform both operations on it at once, before
+loading another cache line.
+```
+for (int i = 0; i < MAXN; i++) {
+    A[i] += j;
+    A[i] *= j;
+}
+```
+Locally, this gives me 660ms for the unfused one and 411ms for the fused one.
+
+Loop fusion is a big deal on CPUs, but it’s an even bigger deal on GPUs (and
+other hardware accelerators). As opposed to say, loop permutation, which
+simply improves data locality, loop fusion can actually reduce the number of
+memory loads needed. For example, in the above, it’s a trivial optimizations
+to then rewrite it as
+
+```
+for (int i = 0; i < MAXN; i++) {
+    int t = A[i];
+    t += j;
+    t *= j;
+    A[i] = t;
+}
+```
+This halving in memory loads can often translate directly to halving of
+runtime in more memory bound systems (like GPUs).
+
+#### Loop Fission/Loop Distribution
+This is the opposite of loop fusion. Although loop fusion is useful if you
+can reduce memory loads, it can be counter-productive to have unrelated
+operations jammed together into a single loop nest. Not only does it
+introduce more memory pressure, it also doesn’t allow optimizations like loop
+permutation to be applied to a single operation at a time.
+
+For example, take a loop like this
+```
+int A[MAXN][MAXN], B[MAXN][MAXN];
+for (int i = 0; i < MAXN; i++) {
+    for (int j=0; j<MAXN; j+=2) {
+        A[i][j] ++;
+        B[j][i] ++;
+    }
+}
+```
+
+
+As seen in the loop permutation section, we’d like to iterate along both A
+and B in row-major order. However, the fact that the operations in A and B
+are in one loop nest doesn’t make it possible to do this for both arrays.
+However, if we split this loop, then we can write it like
+```
+int A[MAXN][MAXN], B[MAXN][MAXN];
+
+for (int i = 0; i < MAXN; i++)
+    for (int j=0; j<MAXN; j+=2)
+        A[i][j]++;
+
+for (int j=0; j<MAXN; j+=2)
+    for (int i = 0; i < MAXN; i++)t
+        B[j][i]++;
+```
+Thus iterating in the optimal order for both arrays.
+
+## Cost Model
+
+For class discussion, think about the following algorithm:
+
+<img src="cost_model.png" style="width: 100%">
+
+They use this cost model for determining the optimal sequence of loop fusion/fission/permutation to apply.
+
+# Future Work
+The loop optimizations presented here are still used everywhere. However,
+these are perhaps the more straightforward optimizations to make. In
+particular, they don't deal with any optimizations (besides loop reversal)
+that change the structure of a loop. For example, unrolling (and the
+vectorization opportunities that poses), as well as tiling.
+
+Incorporating these optimizations in an automated cost model is an ongoing
+problem. Several frameworks for reasoning about these optimizatons have been
+proposed (polyhedral optimization, Halide).
+
+In addition, more cost models have been proposed for handling these
+optimizations as well.
```

Comment:
This is a bit vague—if they've been proposed, can you refer to them specifically?